File: internal/netctl/tc.go
package netctl

import (
	"fmt"
	"log"
	"os/exec"
	"strconv"
	"strings"
)

const uncapped = 1000000 // 1 Gbps in kbit

// TCController manages dual HTB qdisc trees:
//   - WAN interface egress: upload shaping (root class = curve rate)
//   - LAN interface egress: download shaping (uses an intermediate parent class
//     at the curve rate so unmatched local/inter-LAN traffic is not throttled)
//
// This replaces the previous IFB-based approach which had two problems:
// 1. tc ingress redirect ran before nftables, so fw marks were never set on IFB
// 2. Redirecting all WAN ingress broke router-local services
type TCController struct {
	wanIface string
	lanIface string
	minRate  int // minimum rate in kbit
}

// NewTCController creates a tc controller for the given interfaces.
func NewTCController(wanIface, lanIface string, minRateKbit int) *TCController {
	return &TCController{
		wanIface: wanIface,
		lanIface: lanIface,
		minRate:  minRateKbit,
	}
}

// SetupHTB initializes the HTB qdisc trees on WAN and LAN interfaces.
func (tc *TCController) SetupHTB(rootRateKbit int) error {
	if err := tc.setupWAN(rootRateKbit); err != nil {
		return err
	}
	return tc.setupLAN(rootRateKbit)
}

// setupWAN creates the upload shaping tree on the WAN interface.
// Tree structure:
//
//	1: (root HTB, default 2)
//	└── 1:1 (rate=curveRate, ceil=curveRate)
//	    ├── 1:2 (default catch-all, rate=minRate, ceil=curveRate)
//	    └── 1:10+ (per-device classes)
func (tc *TCController) setupWAN(rootRateKbit int) error {
	iface := tc.wanIface
	exec.Command("tc", "qdisc", "del", "dev", iface, "root").Run()

	if err := tc.run("tc", "qdisc", "add", "dev", iface, "root", "handle", "1:", "htb",
		"default", "2"); err != nil {
		return fmt.Errorf("add htb qdisc on %s: %w", iface, err)
	}
	if err := tc.run("tc", "class", "add", "dev", iface, "parent", "1:", "classid", "1:1",
		"htb", "rate", fmt.Sprintf("%dkbit", rootRateKbit), "ceil", fmt.Sprintf("%dkbit", rootRateKbit)); err != nil {
		return fmt.Errorf("add root class on %s: %w", iface, err)
	}
	if err := tc.run("tc", "class", "add", "dev", iface, "parent", "1:1", "classid", "1:2",
		"htb", "rate", fmt.Sprintf("%dkbit", tc.minRate), "ceil", fmt.Sprintf("%dkbit", rootRateKbit)); err != nil {
		return fmt.Errorf("add default class on %s: %w", iface, err)
	}
	if err := tc.run("tc", "qdisc", "add", "dev", iface, "parent", "1:2", "fq_codel"); err != nil {
		log.Printf("warning: fq_codel on %s default: %v", iface, err)
	}
	return nil
}

// setupLAN creates the download shaping tree on the LAN interface.
// Unmatched traffic (router-local, inter-LAN) goes to class 1:2 which is
// effectively unlimited. Marked download traffic goes to device classes
// under 1:3 which is rate-limited to the curve rate.
//
// Tree structure:
//
//	1: (root HTB, default 2)
//	└── 1:1 (rate=1Gbps, ceil=1Gbps)
//	    ├── 1:2 (unmatched/local, rate=900Mbps, ceil=1Gbps)
//	    └── 1:3 (download parent, rate=curveRate, ceil=curveRate)
//	        ├── 1:4 (default shaped, rate=minRate, ceil=curveRate)
//	        └── 1:10+ (per-device classes)
func (tc *TCController) setupLAN(downloadRateKbit int) error {
	iface := tc.lanIface
	exec.Command("tc", "qdisc", "del", "dev", iface, "root").Run()

	if err := tc.run("tc", "qdisc", "add", "dev", iface, "root", "handle", "1:", "htb",
		"default", "2"); err != nil {
		return fmt.Errorf("add htb qdisc on %s: %w", iface, err)
	}
	// Root class: high rate to not throttle unmatched traffic
	if err := tc.run("tc", "class", "add", "dev", iface, "parent", "1:", "classid", "1:1",
		"htb", "rate", fmt.Sprintf("%dkbit", uncapped), "ceil", fmt.Sprintf("%dkbit", uncapped)); err != nil {
		return fmt.Errorf("add root class on %s: %w", iface, err)
	}
	// Default class for unmatched traffic (router-local, inter-LAN): essentially unlimited
	localRate := uncapped - downloadRateKbit
	if localRate < tc.minRate {
		localRate = tc.minRate
	}
	if err := tc.run("tc", "class", "add", "dev", iface, "parent", "1:1", "classid", "1:2",
		"htb", "rate", fmt.Sprintf("%dkbit", localRate), "ceil", fmt.Sprintf("%dkbit", uncapped)); err != nil {
		return fmt.Errorf("add default class on %s: %w", iface, err)
	}
	if err := tc.run("tc", "qdisc", "add", "dev", iface, "parent", "1:2", "fq_codel"); err != nil {
		log.Printf("warning: fq_codel on %s default: %v", iface, err)
	}
	// Download shaping parent: limits aggregate download to curve rate
	if err := tc.run("tc", "class", "add", "dev", iface, "parent", "1:1", "classid", "1:3",
		"htb", "rate", fmt.Sprintf("%dkbit", downloadRateKbit), "ceil", fmt.Sprintf("%dkbit", downloadRateKbit)); err != nil {
		return fmt.Errorf("add download parent on %s: %w", iface, err)
	}
	// Default shaped class (for marked traffic without a specific device class)
	if err := tc.run("tc", "class", "add", "dev", iface, "parent", "1:3", "classid", "1:4",
		"htb", "rate", fmt.Sprintf("%dkbit", tc.minRate), "ceil", fmt.Sprintf("%dkbit", downloadRateKbit)); err != nil {
		return fmt.Errorf("add default shaped class on %s: %w", iface, err)
	}
	if err := tc.run("tc", "qdisc", "add", "dev", iface, "parent", "1:4", "fq_codel"); err != nil {
		log.Printf("warning: fq_codel on %s default shaped: %v", iface, err)
	}
	return nil
}

// UpdateRootRate changes the root class rate on both trees.
func (tc *TCController) UpdateRootRate(rateKbit int) error {
	// WAN: update root class 1:1
	if err := tc.run("tc", "class", "change", "dev", tc.wanIface, "parent", "1:", "classid", "1:1",
		"htb", "rate", fmt.Sprintf("%dkbit", rateKbit), "ceil", fmt.Sprintf("%dkbit", rateKbit)); err != nil {
		return fmt.Errorf("update root rate on %s: %w", tc.wanIface, err)
	}
	// LAN: update download parent class 1:3
	if err := tc.run("tc", "class", "change", "dev", tc.lanIface, "parent", "1:1", "classid", "1:3",
		"htb", "rate", fmt.Sprintf("%dkbit", rateKbit), "ceil", fmt.Sprintf("%dkbit", rateKbit)); err != nil {
		return fmt.Errorf("update download parent on %s: %w", tc.lanIface, err)
	}
	return nil
}

// AddDeviceClass creates a class for a device in both HTB trees.
func (tc *TCController) AddDeviceClass(slot int, rateKbit, ceilKbit int) error {
	classID := fmt.Sprintf("1:%d", 10+slot)
	handle := fmt.Sprintf("%d:", 10+slot)
	mark := 100 + slot

	// WAN: device class under 1:1
	if err := tc.addClassOnIface(tc.wanIface, "1:1", classID, handle, mark, rateKbit, ceilKbit); err != nil {
		return err
	}
	// LAN: device class under 1:3 (download parent)
	if err := tc.addClassOnIface(tc.lanIface, "1:3", classID, handle, mark, rateKbit, ceilKbit); err != nil {
		return err
	}
	return nil
}

func (tc *TCController) addClassOnIface(iface, parent, classID, handle string, mark, rateKbit, ceilKbit int) error {
	if err := tc.run("tc", "class", "add", "dev", iface, "parent", parent, "classid", classID,
		"htb", "rate", fmt.Sprintf("%dkbit", rateKbit), "ceil", fmt.Sprintf("%dkbit", ceilKbit)); err != nil {
		return fmt.Errorf("add device class %s on %s: %w", classID, iface, err)
	}
	if err := tc.run("tc", "qdisc", "add", "dev", iface, "parent", classID, "handle", handle, "fq_codel"); err != nil {
		log.Printf("warning: fq_codel on %s %s: %v", iface, classID, err)
	}
	if err := tc.run("tc", "filter", "add", "dev", iface, "parent", "1:", "protocol", "ip",
		"prio", "1", "handle", strconv.Itoa(mark), "fw", "classid", classID); err != nil {
		return fmt.Errorf("add filter for mark %d on %s: %w", mark, iface, err)
	}
	return nil
}

// RemoveDeviceClass removes a device's class from both HTB trees.
func (tc *TCController) RemoveDeviceClass(slot int) error {
	classID := fmt.Sprintf("1:%d", 10+slot)
	for _, iface := range []string{tc.wanIface, tc.lanIface} {
		exec.Command("tc", "class", "del", "dev", iface, "classid", classID).Run()
	}
	return nil
}

// SetDeviceMode updates both HTB trees based on device mode.
func (tc *TCController) SetDeviceMode(slot int, mode string, fairShareKbit int, burstCeilKbit int, downUpRatio float64) error {
	switch mode {
	case "turbo":
		tc.setClass(tc.wanIface, slot, "1:1", fairShareKbit, uncapped)
		tc.setClass(tc.lanIface, slot, "1:3", fairShareKbit, uncapped)
	case "burst":
		downCeil := int(float64(burstCeilKbit) * downUpRatio)
		upCeil := burstCeilKbit - downCeil
		if downCeil < tc.minRate {
			downCeil = tc.minRate
		}
		if upCeil < tc.minRate {
			upCeil = tc.minRate
		}
		tc.setClass(tc.wanIface, slot, "1:1", fairShareKbit, upCeil)
		tc.setClass(tc.lanIface, slot, "1:3", fairShareKbit, downCeil)
	case "sustained":
		downCeil := int(float64(fairShareKbit) * downUpRatio)
		upCeil := fairShareKbit - downCeil
		if downCeil < tc.minRate {
			downCeil = tc.minRate
		}
		if upCeil < tc.minRate {
			upCeil = tc.minRate
		}
		tc.setClass(tc.wanIface, slot, "1:1", upCeil, upCeil)
		tc.setClass(tc.lanIface, slot, "1:3", downCeil, downCeil)
	}
	return nil
}

func (tc *TCController) setClass(iface string, slot int, parent string, rateKbit, ceilKbit int) {
	classID := fmt.Sprintf("1:%d", 10+slot)
	tc.run("tc", "class", "change", "dev", iface, "parent", parent, "classid", classID,
		"htb", "rate", fmt.Sprintf("%dkbit", rateKbit), "ceil", fmt.Sprintf("%dkbit", ceilKbit))
}

// Teardown removes all tc qdiscs from both interfaces.
func (tc *TCController) Teardown() {
	for _, iface := range []string{tc.wanIface, tc.lanIface} {
		exec.Command("tc", "qdisc", "del", "dev", iface, "root").Run()
	}
}

func (tc *TCController) run(name string, args ...string) error {
	cmd := exec.Command(name, args...)
	out, err := cmd.CombinedOutput()
	if err != nil {
		return fmt.Errorf("%s %s: %s: %w", name, strings.Join(args, " "), string(out), err)
	}
	return nil
}

File: src/netctl/tc.rs
use super::{run_cmd, run_cmd_ignore};
use crate::model::DeviceMode;
use tracing::{info, warn};

const UNCAPPED: i32 = 1_000_000; // 1 Gbps in kbit

/// Manages dual HTB qdisc trees:
///   - WAN interface egress: upload shaping (root class = curve rate)
///   - LAN interface egress: download shaping (uses an intermediate parent class
///     at the curve rate so unmatched local/inter-LAN traffic is not throttled)
pub struct TCController {
    wan_iface: String,
    lan_iface: String,
    min_rate: i32,
}

impl TCController {
    pub fn new(wan_iface: &str, lan_iface: &str, min_rate_kbit: i32) -> Self {
        Self {
            wan_iface: wan_iface.to_string(),
            lan_iface: lan_iface.to_string(),
            min_rate: min_rate_kbit,
        }
    }

    pub fn wan_iface(&self) -> &str {
        &self.wan_iface
    }

    pub fn lan_iface(&self) -> &str {
        &self.lan_iface
    }

    /// Initialize the HTB qdisc trees on WAN and LAN interfaces.
    pub fn setup_htb(&self, root_rate_kbit: i32) -> Result<(), String> {
        self.setup_wan(root_rate_kbit)?;
        self.setup_lan(root_rate_kbit)
    }

    /// WAN tree structure:
    ///   1: (root HTB, default 2)
    ///   └── 1:1 (uncapped parent — per-device classes handle shaping)
    ///       ├── 1:2 (default catch-all)
    ///       └── 1:10+ (per-device classes)
    fn setup_wan(&self, _root_rate_kbit: i32) -> Result<(), String> {
        let iface = &self.wan_iface;
        run_cmd_ignore("tc", &["qdisc", "del", "dev", iface, "root"]);

        self.tc(&[
            "qdisc", "add", "dev", iface, "root", "handle", "1:", "htb", "default", "2",
        ])?;
        self.tc(&[
            "class", "add", "dev", iface, "parent", "1:", "classid", "1:1", "htb",
            "rate", &format!("{UNCAPPED}kbit"),
            "ceil", &format!("{UNCAPPED}kbit"),
        ])?;
        self.tc(&[
            "class", "add", "dev", iface, "parent", "1:1", "classid", "1:2", "htb",
            "rate", &format!("{}kbit", self.min_rate),
            "ceil", &format!("{UNCAPPED}kbit"),
        ])?;
        if let Err(e) = self.tc(&["qdisc", "add", "dev", iface, "parent", "1:2", "fq_codel"]) {
            warn!("fq_codel on {iface} default: {e}");
        }
        Ok(())
    }

    /// LAN tree structure:
    ///   1: (root HTB, default 2)
    ///   └── 1:1 (uncapped)
    ///       ├── 1:2 (unmatched/local, uncapped)
    ///       └── 1:3 (download parent, uncapped — per-device classes handle shaping)
    ///           ├── 1:4 (default shaped)
    ///           └── 1:10+ (per-device classes)
    fn setup_lan(&self, _download_rate_kbit: i32) -> Result<(), String> {
        let iface = &self.lan_iface;
        run_cmd_ignore("tc", &["qdisc", "del", "dev", iface, "root"]);

        self.tc(&[
            "qdisc", "add", "dev", iface, "root", "handle", "1:", "htb", "default", "2",
        ])?;
        self.tc(&[
            "class", "add", "dev", iface, "parent", "1:", "classid", "1:1", "htb",
            "rate", &format!("{UNCAPPED}kbit"),
            "ceil", &format!("{UNCAPPED}kbit"),
        ])?;
        self.tc(&[
            "class", "add", "dev", iface, "parent", "1:1", "classid", "1:2", "htb",
            "rate", &format!("{UNCAPPED}kbit"),
            "ceil", &format!("{UNCAPPED}kbit"),
        ])?;
        if let Err(e) = self.tc(&["qdisc", "add", "dev", iface, "parent", "1:2", "fq_codel"]) {
            warn!("fq_codel on {iface} default: {e}");
        }
        self.tc(&[
            "class", "add", "dev", iface, "parent", "1:1", "classid", "1:3", "htb",
            "rate", &format!("{UNCAPPED}kbit"),
            "ceil", &format!("{UNCAPPED}kbit"),
        ])?;
        self.tc(&[
            "class", "add", "dev", iface, "parent", "1:3", "classid", "1:4", "htb",
            "rate", &format!("{}kbit", self.min_rate),
            "ceil", &format!("{UNCAPPED}kbit"),
        ])?;
        if let Err(e) = self.tc(&[
            "qdisc", "add", "dev", iface, "parent", "1:4", "fq_codel",
        ]) {
            warn!("fq_codel on {iface} default shaped: {e}");
        }
        Ok(())
    }

    /// No-op — parent classes are always uncapped.
    /// Per-device classes handle all rate limiting.
    pub fn update_root_rate(&self, _rate_kbit: i32, _has_turbo: bool) -> Result<(), String> {
        Ok(())
    }

    /// Check if the HTB root qdisc exists on the WAN interface.
    pub fn htb_exists(&self) -> bool {
        if let Ok(output) = run_cmd("tc", &["qdisc", "show", "dev", &self.wan_iface]) {
            output.contains("htb")
        } else {
            false
        }
    }

    /// Re-create HTB trees only if they are missing.
    pub fn ensure_htb(&self, rate_kbit: i32) -> Result<(), String> {
        if !self.htb_exists() {
            info!("HTB trees missing, re-creating");
            self.setup_htb(rate_kbit)?;
        }
        Ok(())
    }

    /// Create a class for a device in both HTB trees.
    pub fn add_device_class(
        &self,
        slot: i32,
        rate_kbit: i32,
        ceil_kbit: i32,
    ) -> Result<(), String> {
        let class_id = format!("1:{}", 10 + slot);
        let handle = format!("{}:", 10 + slot);
        let mark = (100 + slot).to_string();

        // rate == ceil: Rust state machine controls mode, HTB enforces hard cap
        let rate = ceil_kbit.max(rate_kbit).max(1);
        // WAN: device class under 1:1
        self.add_class_on_iface(&self.wan_iface, "1:1", &class_id, &handle, &mark, rate, rate)?;
        // LAN: device class under 1:3 (download parent)
        self.add_class_on_iface(&self.lan_iface, "1:3", &class_id, &handle, &mark, rate, rate)?;
        Ok(())
    }

    #[allow(clippy::too_many_arguments)]
    fn add_class_on_iface(
        &self,
        iface: &str,
        parent: &str,
        class_id: &str,
        handle: &str,
        mark: &str,
        rate_kbit: i32,
        ceil_kbit: i32,
    ) -> Result<(), String> {
        self.tc(&[
            "class", "add", "dev", iface, "parent", parent, "classid", class_id, "htb", "rate",
            &format!("{rate_kbit}kbit"),
            "ceil",
            &format!("{ceil_kbit}kbit"),
        ])?;
        if let Err(e) = self.tc(&[
            "qdisc", "add", "dev", iface, "parent", class_id, "handle", handle, "fq_codel",
        ]) {
            warn!("fq_codel on {iface} {class_id}: {e}");
        }
        self.tc(&[
            "filter", "add", "dev", iface, "parent", "1:", "protocol", "ip", "prio", "1",
            "handle", mark, "fw", "classid", class_id,
        ])?;
        Ok(())
    }

    /// Remove a device's class from both HTB trees.
    pub fn remove_device_class(&self, slot: i32) {
        let class_id = format!("1:{}", 10 + slot);
        for iface in [&self.wan_iface, &self.lan_iface] {
            run_cmd_ignore("tc", &["class", "del", "dev", iface, "classid", &class_id]);
        }
    }

    /// Update both HTB trees based on device mode.
    /// rate == ceil everywhere — Rust state machine controls mode transitions,
    /// HTB just enforces the current hard cap.
    pub fn set_device_mode(
        &self,
        slot: i32,
        mode: DeviceMode,
        fair_share_kbit: i32,
        burst_ceil_kbit: i32,
        down_up_ratio: f64,
    ) {
        match mode {
            DeviceMode::Turbo => {
                self.set_class(&self.wan_iface, slot, "1:1", UNCAPPED, UNCAPPED);
                self.set_class(&self.lan_iface, slot, "1:3", UNCAPPED, UNCAPPED);
            }
            DeviceMode::Burst => {
                let down = (burst_ceil_kbit as f64 * down_up_ratio) as i32;
                let up = burst_ceil_kbit - down;
                self.set_class(&self.wan_iface, slot, "1:1", up.max(1), up.max(1));
                self.set_class(&self.lan_iface, slot, "1:3", down.max(1), down.max(1));
            }
            DeviceMode::Sustained => {
                let down = (fair_share_kbit as f64 * down_up_ratio) as i32;
                let up = fair_share_kbit - down;
                self.set_class(&self.wan_iface, slot, "1:1", up.max(1), up.max(1));
                self.set_class(&self.lan_iface, slot, "1:3", down.max(1), down.max(1));
            }
        }
    }

    fn set_class(&self, iface: &str, slot: i32, parent: &str, rate_kbit: i32, ceil_kbit: i32) {
        let class_id = format!("1:{}", 10 + slot);
        let rate = format!("{rate_kbit}kbit");
        let ceil = format!("{ceil_kbit}kbit");

        // Try "change" first; if the class was lost (e.g., qdisc reset),
        // fall back to "add" so the device doesn't run unshaped.
        if self
            .tc(&[
                "class", "change", "dev", iface, "parent", parent, "classid", &class_id, "htb",
                "rate", &rate, "ceil", &ceil,
            ])
            .is_err()
        {
            warn!(
                "tc class change failed for {class_id} on {iface}, re-adding"
            );
            if let Err(e) = self.tc(&[
                "class", "add", "dev", iface, "parent", parent, "classid", &class_id, "htb",
                "rate", &rate, "ceil", &ceil,
            ]) {
                warn!("tc class add also failed for {class_id} on {iface}: {e}");
                return;
            }
            // Re-add fq_codel and filter for the recreated class
            let handle = format!("{}:", 10 + slot);
            let mark = (100 + slot).to_string();
            let _ = self.tc(&[
                "qdisc", "add", "dev", iface, "parent", &class_id, "handle", &handle, "fq_codel",
            ]);
            let _ = self.tc(&[
                "filter", "add", "dev", iface, "parent", "1:", "protocol", "ip", "prio", "1",
                "handle", &mark, "fw", "classid", &class_id,
            ]);
        }
    }

    /// Remove all tc qdiscs from both interfaces.
    pub fn teardown(&self) {
        for iface in [&self.wan_iface, &self.lan_iface] {
            run_cmd_ignore("tc", &["qdisc", "del", "dev", iface, "root"]);
        }
    }

    fn tc(&self, args: &[&str]) -> Result<(), String> {
        run_cmd("tc", args).map(|_| ())
    }
}
